File: Components/WinForms/VB/EditorsMainDemo/Modules/ModuleBreadCrumbEdit.vb
Imports System
Imports System.IO
Imports System.Windows.Forms
Imports DevExpress.Utils.Helpers
Imports DevExpress.XtraEditors.Repository

Namespace DevExpress.XtraEditors.Demos
	Partial Public Class ModuleBreadCrumbEdit
		Inherits TutorialControl
		Private currentPath As String
		Private properties As RepositoryItemBreadCrumbEdit
		Public Sub New()
            InitializeComponent()
            Me.properties = BreadCrumb.Properties
            TutorialInfo.WhatsThisCodeFile = "VB\EditorsMainDemo\Modules\ModuleBreadCrumbEdit.vb"
            TutorialInfo.WhatsThisXMLFile = "breadCrumbEdit.xml"
		End Sub
		Protected Overrides Sub OnLoad(ByVal e As EventArgs)
			MyBase.OnLoad(e)
			InitValues()
		End Sub
		Private Sub InitValues()
			Me.currentPath = Environment.GetFolderPath(Environment.SpecialFolder.Desktop)
			BreadCrumb.Path = Me.currentPath
			For Each driveInfo As DriveInfo In FileSystemHelper.GetFixedDrives()
				BreadCrumb.Properties.History.Add(New BreadCrumbHistoryItem(driveInfo.RootDirectory.ToString()))
			Next driveInfo
			Me.cbeMode.Properties.Items.AddEnum(GetType(BreadCrumbMode))
			Me.cbeMode.EditValue = Me.properties.BreadCrumbMode
			Me.seDropDownRowCount.Value = Me.properties.NodeDropDownRowCount
			Me.ceShowRootGlyph.Checked = Me.properties.ShowRootGlyph
			If Me.cbeImages.Properties.Items.Count > 0 Then
				Me.cbeImages.SelectedIndex = 0
			End If
		End Sub
		' <breadCrumbEdit>
		Private Sub OnBreadCrumbPathChanged(ByVal sender As Object, ByVal e As BreadCrumbPathChangedEventArgs) Handles breadCrumbEdit.Properties.PathChanged
			Me.pathLabel.Text = e.Path
			Me.currentPath = Me.pathLabel.Text
			AddEventEntry("PathChanged", e.Path)
			UpdateControls()
		End Sub
		Private Sub OnBreadCrumbNewNodeAdding(ByVal sender As Object, ByVal e As BreadCrumbNewNodeAddingEventArgs) Handles breadCrumbEdit.Properties.NewNodeAdding
			AddEventEntry("NewNodeAdding", e.Node)
			e.Node.PopulateOnDemand = True
		End Sub
		Private Sub OnBreadCrumbQueryChildNodes(ByVal sender As Object, ByVal e As BreadCrumbQueryChildNodesEventArgs) Handles breadCrumbEdit.Properties.QueryChildNodes
			AddEventEntry("QueryChildNodes", e.Node)
			If e.Node.Caption = "Root" Then
				InitBreadCrumbRootNode(e.Node)
				Return
			End If
			If e.Node.Caption = "Computer" Then
				InitBreadCrumbComputerNode(e.Node)
				Return
			End If
			Dim dir As String = e.Node.Path
			If Not FileSystemHelper.IsDirExists(dir) Then
				Return
			End If
			Dim subDirs() As String = FileSystemHelper.GetSubFolders(dir)
			For i As Integer = 0 To subDirs.Length - 1
				e.Node.ChildNodes.Add(CreateNode(subDirs(i)))
			Next i
		End Sub
		Private Sub OnBreadCrumbValidatePath(ByVal sender As Object, ByVal e As BreadCrumbValidatePathEventArgs) Handles breadCrumbEdit.Properties.ValidatePath
			AddEventEntry("ValidatePath", e.Path)
			If Not FileSystemHelper.IsDirExists(e.Path) Then
				e.ValidationResult = BreadCrumbValidatePathResult.Cancel
				Return
			End If
			e.ValidationResult = BreadCrumbValidatePathResult.CreateNodes
		End Sub
		Private Sub OnBreadCrumbRootGlyphClick(ByVal sender As Object, ByVal e As EventArgs) Handles breadCrumbEdit.Properties.RootGlyphClick
			AddEventEntry("RootGlyphClick", sender)
			BreadCrumb.Properties.BreadCrumbMode = BreadCrumbMode.Edit
			BreadCrumb.SelectAll()
		End Sub
		Private Sub OnBreadCrumbNodeClick(ByVal sender As Object, ByVal e As BreadCrumbNodeClickEventArgs) Handles breadCrumbEdit.Properties.NodeClick
			AddEventEntry("NodeClick", e.Node)
		End Sub
		Private Sub OnBreadCrumbShowUserActionMenu(ByVal sender As Object, ByVal e As BreadCrumbShowUserActionMenuEventArgs) Handles breadCrumbEdit.Properties.ShowUserActionMenu
			AddEventEntry("ShowUserActionMenu", sender)
		End Sub
		Private Sub OnBreadCrumbShownNodeDropdown(ByVal sender As Object, ByVal e As BreadCrumbShownNodeDropDownEventArgs) Handles breadCrumbEdit.Properties.ShownNodeDropdown
			AddEventEntry("ShownNodeDropdown", e.Node)
		End Sub
		Private Sub OnBreadCrumbHiddenNodeDropDown(ByVal sender As Object, ByVal e As BreadCrumbHiddenNodeDropDownEventArgs) Handles breadCrumbEdit.Properties.HiddenNodeDropDown
			AddEventEntry("HiddenNodeDropDown", e.Node)
		End Sub
		Private Sub OnBreadCrumbPathRejected(ByVal sender As Object, ByVal e As BreadCrumbPathRejectedEventArgs) Handles breadCrumbEdit.Properties.PathRejected
			AddEventEntry("PathRejected", e.Path)
		End Sub
		Private Sub OnBreadCrumbSelectedNodeChanged(ByVal sender As Object, ByVal e As BreadCrumbSelectedNodeChangedEventArgs) Handles breadCrumbEdit.Properties.SelectedNodeChanged
			AddEventEntry("SelectedNodeChanged", e.NewNode)
			Me.selNodeValueLabel.Text = If(e.NewNode IsNot Nothing, e.NewNode.ToString(), "(none)")
		End Sub
		Private Sub OnBreadCrumbSelectorClientEmptySpaceClick(ByVal sender As Object, ByVal e As MouseEventArgs) Handles breadCrumbEdit.Properties.SelectorClientEmptySpaceClick
			AddEventEntry("SelectorClientEmptySpaceClick", e.Location)
		End Sub
		Private Sub InitBreadCrumbRootNode(ByVal node As BreadCrumbNode)
			node.ChildNodes.Add(New BreadCrumbNode("Desktop", Environment.GetFolderPath(Environment.SpecialFolder.Desktop)))
			node.ChildNodes.Add(New BreadCrumbNode("Documents", Environment.GetFolderPath(Environment.SpecialFolder.Recent)))
			node.ChildNodes.Add(New BreadCrumbNode("Music", Environment.GetFolderPath(Environment.SpecialFolder.MyMusic)))
			node.ChildNodes.Add(New BreadCrumbNode("Pictures", Environment.GetFolderPath(Environment.SpecialFolder.MyPictures)))
			node.ChildNodes.Add(New BreadCrumbNode("Video", Environment.GetFolderPath(Environment.SpecialFolder.MyVideos)))
			node.ChildNodes.Add(New BreadCrumbNode("Program Files", Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles)))
			node.ChildNodes.Add(New BreadCrumbNode("Windows", Environment.GetFolderPath(Environment.SpecialFolder.Windows)))
		End Sub
		Private Sub InitBreadCrumbComputerNode(ByVal node As BreadCrumbNode)
			For Each driveInfo As DriveInfo In FileSystemHelper.GetFixedDrives()
				node.ChildNodes.Add(New BreadCrumbNode(driveInfo.Name, driveInfo.RootDirectory))
			Next driveInfo
		End Sub
		Private Function CreateNode(ByVal path As String) As BreadCrumbNode
			Dim folderName As String = FileSystemHelper.GetDirName(path)
			Return New BreadCrumbNode(folderName, folderName, True)
		End Function
		' </breadCrumbEdit>
		' <btnClearEventLog>
		Private Sub OnClearEventLogClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnClearEventLog.Click
			Me.eventsLog.Text = String.Empty
		End Sub
		' </btnClearEventLog>
		' <cbeMode>
		Private Sub OnBreadCrumbModeChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbeMode.SelectedIndexChanged
			Dim comboBox As ComboBoxEdit = CType(sender, ComboBoxEdit)
			Me.properties.BreadCrumbMode = CType(comboBox.EditValue, BreadCrumbMode)
		End Sub
		' </cbeMode>
		' <seDropDownRowCount>
		Private Sub OnBreadCrumbDropDownRowCountChanged(ByVal sender As Object, ByVal e As EventArgs) Handles seDropDownRowCount.EditValueChanged
            If Me.properties Is Nothing Then
                Return
            End If
            Dim spinEdit As SpinEdit = CType(sender, SpinEdit)
            Me.properties.NodeDropDownRowCount = CInt(Fix(spinEdit.Value))
        End Sub
		' </seDropDownRowCount>
		' <ceShowRootGlyph>
		Private Sub OnShowRootGlyphCheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles ceShowRootGlyph.CheckedChanged
			Dim checkEdit As CheckEdit = CType(sender, CheckEdit)
			Me.properties.ShowRootGlyph = checkEdit.Checked
		End Sub
		' </ceShowRootGlyph>
		' <cbeImages>
		Private Sub OnBreadCrumbSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbeImages.SelectedIndexChanged
			Dim comboBox As ComboBoxEdit = CType(sender, ComboBoxEdit)
			Me.properties.ImageIndex = comboBox.SelectedIndex
			Me.properties.RootImageIndex = Me.properties.ImageIndex
		End Sub
		' </cbeImages>
		' <btnGoUp>
		Private Sub OnGoUpClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnGoUp.Click
            Me.BreadCrumb.GoUp()
        End Sub
		' </btnGoUp>
		' <btnGoBack>
		Private Sub OnGoBackClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnGoBack.Click
            Me.BreadCrumb.GoBack()
        End Sub
		' </btnGoBack>
		' <btnGoForward>
		Private Sub OnGoForwardClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnGoForward.Click
            Me.BreadCrumb.GoForward()
        End Sub
		' </btnGoForward>

		Private Sub AddEventEntry(ByVal entry As String, ByVal obj As Object)
			Dim val As String = String.Empty
			If obj IsNot Nothing Then
				val = If(TypeOf obj Is BreadCrumbEdit, GetType(BreadCrumbEdit).Name, obj.ToString())
			Else
				val = "(null)"
			End If
			Me.eventsLog.Text = String.Format("{0}: {1}{2}", entry, val, Environment.NewLine) + eventsLog.Text
		End Sub
		Private Sub UpdateControls()
            Me.btnGoBack.Enabled = Me.BreadCrumb.CanGoBack
            Me.btnGoForward.Enabled = Me.BreadCrumb.CanGoForward
            Me.btnGoUp.Enabled = Me.BreadCrumb.CanGoUp
        End Sub
        Private ReadOnly Property BreadCrumb() As BreadCrumbEdit
            Get
                Return breadCrumbEdit
            End Get
        End Property
	End Class
End Namespace

File: Components/WinForms/CS/EditorsMainDemo/Modules/ModuleBreadCrumbEdit.cs
using System;
using System.IO;
using System.Windows.Forms;
using DevExpress.Utils.Helpers;
using DevExpress.XtraEditors.Repository;

namespace DevExpress.XtraEditors.Demos {
    public partial class ModuleBreadCrumbEdit : TutorialControl {
        string currentPath;
        RepositoryItemBreadCrumbEdit properties;
        public ModuleBreadCrumbEdit() {
            InitializeComponent();
            this.properties = BreadCrumb.Properties;
            TutorialInfo.WhatsThisCodeFile = "CS\\EditorsMainDemo\\Modules\\ModuleBreadCrumbEdit.cs";
            TutorialInfo.WhatsThisXMLFile = "DevExpress.XtraEditors.Demos.CodeInfo.breadCrumbEdit.xml";
        }
        protected override void UpdateTabControlPosition(XtraTab.XtraTabControl tabControl) { }
        protected override void CalcContentCore() {
            CalcLayoutControlBounds(layoutControl1);
        }
        protected override void OnLoad(EventArgs e) {
            base.OnLoad(e);
            InitValues();
        }
        void InitValues() {
            BreadCrumb.Path = this.currentPath = Environment.GetFolderPath(Environment.SpecialFolder.Desktop);
            foreach(DriveInfo driveInfo in FileSystemHelper.GetFixedDrives()) {
                BreadCrumb.Properties.History.Add(new BreadCrumbHistoryItem(driveInfo.RootDirectory.ToString()));
            }
            this.cbeMode.Properties.Items.AddEnum(typeof(BreadCrumbMode));
            this.cbeMode.EditValue = this.properties.BreadCrumbMode;
            this.seDropDownRowCount.Value = this.properties.NodeDropDownRowCount;
            this.ceShowRootGlyph.Checked = this.properties.ShowRootGlyph;
            if(this.cbeImages.Properties.Items.Count > 0) this.cbeImages.SelectedIndex = 0;
        }
        // <breadCrumbEdit>
        void OnBreadCrumbPathChanged(object sender, BreadCrumbPathChangedEventArgs e) {
            this.currentPath = this.pathLabel.Text = e.Path;
            AddEventEntry("PathChanged", e.Path);
            UpdateControls();
        }
        void OnBreadCrumbNewNodeAdding(object sender, BreadCrumbNewNodeAddingEventArgs e) {
            AddEventEntry("NewNodeAdding", e.Node);
            e.Node.PopulateOnDemand = true;
        }
        void OnBreadCrumbQueryChildNodes(object sender, BreadCrumbQueryChildNodesEventArgs e) {
            AddEventEntry("QueryChildNodes", e.Node);
            if(e.Node.Caption == "Root") {
                InitBreadCrumbRootNode(e.Node);
                return;
            }
            if(e.Node.Caption == "Computer") {
                InitBreadCrumbComputerNode(e.Node);
                return;
            }
            string dir = e.Node.Path;
            if(!FileSystemHelper.IsDirExists(dir))
                return;
            string[] subDirs = FileSystemHelper.GetSubFolders(dir);
            for(int i = 0; i < subDirs.Length; i++) {
                e.Node.ChildNodes.Add(CreateNode(subDirs[i]));
            }
        }
        void OnBreadCrumbValidatePath(object sender, BreadCrumbValidatePathEventArgs e) {
            AddEventEntry("ValidatePath", e.Path);
            if(!FileSystemHelper.IsDirExists(e.Path)) {
                e.ValidationResult = BreadCrumbValidatePathResult.Cancel;
                return;
            }
            e.ValidationResult = BreadCrumbValidatePathResult.CreateNodes;
        }
        void OnBreadCrumbRootGlyphClick(object sender, EventArgs e) {
            AddEventEntry("RootGlyphClick", sender);
            BreadCrumb.Properties.BreadCrumbMode = BreadCrumbMode.Edit;
            BreadCrumb.SelectAll();
        }
        void OnBreadCrumbNodeClick(object sender, BreadCrumbNodeClickEventArgs e) {
            AddEventEntry("NodeClick", e.Node);
        }
        void OnBreadCrumbShowUserActionMenu(object sender, BreadCrumbShowUserActionMenuEventArgs e) {
            AddEventEntry("ShowUserActionMenu", sender);
        }
        void OnBreadCrumbShownNodeDropdown(object sender, BreadCrumbShownNodeDropDownEventArgs e) {
            AddEventEntry("ShownNodeDropdown", e.Node);
        }
        void OnBreadCrumbHiddenNodeDropDown(object sender, BreadCrumbHiddenNodeDropDownEventArgs e) {
            AddEventEntry("HiddenNodeDropDown", e.Node);
        }
        void OnBreadCrumbPathRejected(object sender, BreadCrumbPathRejectedEventArgs e) {
            AddEventEntry("PathRejected", e.Path);
        }
        void OnBreadCrumbSelectedNodeChanged(object sender, BreadCrumbSelectedNodeChangedEventArgs e) {
            AddEventEntry("SelectedNodeChanged", e.NewNode);
            this.selNodeValueLabel.Text = e.NewNode != null ? e.NewNode.ToString() : "(none)";
        }
        void OnBreadCrumbSelectorClientEmptySpaceClick(object sender, MouseEventArgs e) {
            AddEventEntry("SelectorClientEmptySpaceClick", e.Location);
        }
        void InitBreadCrumbRootNode(BreadCrumbNode node) {
            node.ChildNodes.Add(new BreadCrumbNode("Desktop", Environment.GetFolderPath(Environment.SpecialFolder.Desktop)));
            node.ChildNodes.Add(new BreadCrumbNode("Documents", Environment.GetFolderPath(Environment.SpecialFolder.Recent)));
            node.ChildNodes.Add(new BreadCrumbNode("Music", Environment.GetFolderPath(Environment.SpecialFolder.MyMusic)));
            node.ChildNodes.Add(new BreadCrumbNode("Pictures", Environment.GetFolderPath(Environment.SpecialFolder.MyPictures)));
            node.ChildNodes.Add(new BreadCrumbNode("Video", Environment.GetFolderPath(Environment.SpecialFolder.MyVideos)));
            node.ChildNodes.Add(new BreadCrumbNode("Program Files", Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles)));
            node.ChildNodes.Add(new BreadCrumbNode("Windows", Environment.GetFolderPath(Environment.SpecialFolder.Windows)));
        }
        void InitBreadCrumbComputerNode(BreadCrumbNode node) {
            foreach(DriveInfo driveInfo in FileSystemHelper.GetFixedDrives()) {
                node.ChildNodes.Add(new BreadCrumbNode(driveInfo.Name, driveInfo.RootDirectory));
            }
        }
        BreadCrumbNode CreateNode(string path) {
            string folderName = FileSystemHelper.GetDirName(path);
            return new BreadCrumbNode(folderName, folderName, true);
        }
        // </breadCrumbEdit>
        // <btnClearEventLog>
        void OnClearEventLogClick(object sender, EventArgs e) {
            this.eventsLog.Text = string.Empty;
        }
        // </btnClearEventLog>
        // <cbeMode>
        void OnBreadCrumbModeChanged(object sender, EventArgs e) {
            ComboBoxEdit comboBox = (ComboBoxEdit)sender;
            this.properties.BreadCrumbMode = (BreadCrumbMode)comboBox.EditValue;
        }
        // </cbeMode>
        // <seDropDownRowCount>
        void OnBreadCrumbDropDownRowCountChanged(object sender, EventArgs e) {
            SpinEdit spinEdit = (SpinEdit)sender;
            this.properties.NodeDropDownRowCount = (int)spinEdit.Value;
        }
        // </seDropDownRowCount>
        // <ceShowRootGlyph>
        void OnShowRootGlyphCheckedChanged(object sender, EventArgs e) {
            CheckEdit checkEdit = (CheckEdit)sender;
            this.properties.ShowRootGlyph = checkEdit.Checked;
        }
        // </ceShowRootGlyph>
        // <cbeImages>
        void OnBreadCrumbSelectedIndexChanged(object sender, EventArgs e) {
            ComboBoxEdit comboBox = (ComboBoxEdit)sender;
            this.properties.RootImageIndex = this.properties.ImageIndex = comboBox.SelectedIndex;
        }
        // </cbeImages>
        // <btnGoUp>
        void OnGoUpClick(object sender, EventArgs e) {
            BreadCrumb.GoUp();
        }
        // </btnGoUp>
        // <btnGoBack>
        void OnGoBackClick(object sender, EventArgs e) {
            BreadCrumb.GoBack();
        }
        // </btnGoBack>
        // <btnGoForward>
        void OnGoForwardClick(object sender, EventArgs e) {
            BreadCrumb.GoForward();
        }
        // </btnGoForward>

        void AddEventEntry(string entry, object obj) {
            string val = string.Empty;
            if(obj != null) {
                val = (obj is BreadCrumbEdit) ? typeof(BreadCrumbEdit).Name : obj.ToString();
            }
            else {
                val = "(null)";
            }
            this.eventsLog.Text = string.Format("{0}: {1}{2}", entry, val, Environment.NewLine) + eventsLog.Text;
        }
        void UpdateControls() {
            this.btnGoBack.Enabled = BreadCrumb.CanGoBack;
            this.btnGoForward.Enabled = BreadCrumb.CanGoForward;
            this.btnGoUp.Enabled = BreadCrumb.CanGoUp;
        }
        BreadCrumbEdit BreadCrumb { get { return breadCrumbEdit; } }
    }
}
